sql/src/main/java/io/crate/planner/operators/InsertFromValues.java

```diff
@@ -0,0 +1,766 @@
+                    tableInfo.ident(),
+                    "Invalid value '" + value + "' for type '" + targetType + "'");
+            }
+        }
+        return new RowN(cells);
+    }
+
+    private static ShardLocation getShardLocation(String indexName,
+                                                  String id,
+                                                  @Nullable String routing,
+                                                  ClusterService clusterService) {
+        ShardIterator shardIterator = clusterService.operationRouting().indexShards(
+            clusterService.state(),
+            indexName,
+            id,
+            routing);
+
+        final String nodeId;
+        ShardRouting shardRouting = shardIterator.nextOrNull();
+        if (shardRouting == null) {
+            nodeId = null;
+        } else if (shardRouting.active() == false) {
+            nodeId = shardRouting.relocatingNodeId();
+        } else {
+            nodeId = shardRouting.currentNodeId();
+        }
+        return new ShardLocation(shardIterator.shardId(), nodeId);
+    }
+
+    private static <TReq extends ShardRequest<TReq, TItem>, TItem extends ShardRequest.Item>
+        Map<ShardLocation, TReq> resolveAndGroupShardRequests(ShardedRequests<TReq, TItem> shardedRequests,
+                                                          ClusterService clusterService) {
+        var itemsByMissingIndex = shardedRequests.itemsByMissingIndex().entrySet().iterator();
+        while (itemsByMissingIndex.hasNext()) {
+            var entry = itemsByMissingIndex.next();
+            var index = entry.getKey();
+            var requestItems = entry.getValue();
+
+            var requestItemsIterator = requestItems.iterator();
+            while (requestItemsIterator.hasNext()) {
+                var itemAndRoutingAndSourceInfo = requestItemsIterator.next();
+                ShardLocation shardLocation;
+                try {
+                    shardLocation = getShardLocation(
+                        index,
+                        itemAndRoutingAndSourceInfo.item().id(),
+                        itemAndRoutingAndSourceInfo.routing(),
+                        clusterService);
+                } catch (IndexNotFoundException e) {
+                    if (IndexParts.isPartitioned(index)) {
+                        requestItemsIterator.remove();
+                        continue;
+                    } else {
+                        throw e;
+                    }
+                }
+                shardedRequests.add(itemAndRoutingAndSourceInfo.item(), 0, shardLocation, null);
+                requestItemsIterator.remove();
+            }
+            if (requestItems.isEmpty()) {
+                itemsByMissingIndex.remove();
+            }
+        }
+
+        return shardedRequests.itemsByShard();
+    }
+
+    private CompletableFuture<ShardResponse.CompressedResult> execute(Collection<ShardUpsertRequest> shardUpsertRequests,
```

Comment:
requests execution and results parsing are taken from legacy shard upsert logic.